src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -0,0 +1,373 @@
+/**
+ * @file methods/ann/layer/batch_norm_impl.hpp
+ * @author Praveen Ch
+ * @author Manthan-R-Sheth
+ * @author Kartik Dutt
+ * @author Shubham Agrawal
+ *
+ * Implementation of the Batch Normalization Layer.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LAYER_BATCHNORM_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_BATCHNORM_IMPL_HPP
+
+// In case it is not included.
+#include "batch_norm.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename MatType>
+BatchNormType<MatType>::BatchNormType() : 
+    Layer<MatType>(),
+    maxAxis(2),
+    eps(1e-8),
+    average(true),
+    momentum(0.0),
+    count(0),
+    averageFactor(0.0),
+    inputDimension(1),
+    size(0),
+    higherDimension(1)
+{
+  // Nothing to do here.
+}
+
+template <typename MatType>
+BatchNormType<MatType>::BatchNormType(
+    const size_t maxAxis,
+    const double eps,
+    const bool average,
+    const double momentum) : 
+    Layer<MatType>(),
+    maxAxis(maxAxis),
+    eps(eps),
+    average(average),
+    momentum(momentum),
+    count(0),
+    averageFactor(0.0),
+    inputDimension(1),
+    size(0),
+    higherDimension(1)
+{
+  // Nothing to do here.
+}
+
+// Copy constructor.
+template<typename MatType>
+BatchNormType<MatType>::BatchNormType(const BatchNormType& layer) :
+    Layer<MatType>(layer),
+    maxAxis(layer.maxAxis),
+    eps(layer.eps),
+    average(layer.average),
+    momentum(layer.momentum),
+    count(layer.count),
+    averageFactor(layer.averageFactor),
+    inputDimension(layer.inputDimension),
+    size(layer.size),
+    higherDimension(layer.higherDimension),
+    runningMean(layer.runningMean),
+    runningVariance(layer.runningVariance)
```

Comment:
What about `mean`, `variance`, and other members? Should we copy those too?

Reply:
I think there won't be too much extra overhead. That's why I removed it.

Reply:
Yep, sounds good to me. :+1: